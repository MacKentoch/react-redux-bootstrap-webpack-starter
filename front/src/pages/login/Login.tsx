import React, { useState, useEffect, useCallback } from 'react';
import { RouteComponentProps } from 'react-router';
import Button from 'reactstrap/lib/Button';
import Row from 'reactstrap/lib/Row';
import Col from 'reactstrap/lib/Col';
import auth from '../../services/auth';
import FadeInEntrance from '../../components/fadeInEntrance';
import { MappedDispatchToProps, MappedStateToProps, OwnProps } from './index';

// #region types
export type Props = RouteComponentProps &
  MappedStateToProps &
  MappedDispatchToProps &
  OwnProps;
// #endregion

function Login({
  location,
  isLogging,
  isFetching,
  disconnectUser,
  logUserIfNeeded,
  history,
}: Props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  useEffect(() => {
    // on mount disconnect (clear previous user auth and info):
    disconnectUser();
  }, []);

  // #region form inputs change callbacks
  const handlesOnEmailChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (event) {
        event.preventDefault();
        setEmail(event.target.value.trim());
      }
    },
    [],
  );

  const handlesOnPasswordChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (event) {
        event.preventDefault();
        setPassword(event.target.value.trim());
      }
    },
    [],
  );
  // #endregion

  // #region on login button click callback
  const handlesOnLogin = useCallback(
    async (event: React.MouseEvent<HTMLButtonElement>) => {
      event && event.preventDefault();

      try {
        const response = await logUserIfNeeded(email, password);
        const {
          data: { token, user },
        } = response.payload;

        auth.setToken(token);
        auth.setUserInfo(user);

        // @ts-ignore
        const { from } = location.state || { from: { pathname: '/' } };

        history.replace(from); // back to Home
      } catch (error) {
        /* eslint-disable no-console */
        console.log('login went wrong..., error: ', error);
        /* eslint-enable no-console */
      }
    },
    [history, location, email, password],
  );
  // #endregion

  // #region on go back home button click callback
  const goHome = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      event && event.preventDefault();
      history.push({ pathname: '/' });
    },
    [history],
  );
  // #endregion

  return (
    <FadeInEntrance>
      <div className="content">
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <form className="form-horizontal" noValidate>
              <fieldset>
                <legend>Login</legend>
                <div className="form-group">
                  <label
                    htmlFor="inputEmail"
                    className="col-lg-2 control-label"
                  >
                    Email
                  </label>
                  <Col lg={12}>
                    <input
                      type="text"
                      className="form-control"
                      id="inputEmail"
                      placeholder="Email"
                      autoComplete="username email"
                      value={email}
                      onChange={handlesOnEmailChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <label
                    htmlFor="inputPassword"
                    className="col-lg-2 control-label"
                  >
                    Password
                  </label>
                  <Col lg={12}>
                    <input
                      type="password"
                      className="form-control"
                      id="inputPassword"
                      placeholder="Password"
                      autoComplete="current-password"
                      value={password}
                      onChange={handlesOnPasswordChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <Col lg={{ size: 12 }}>
                    <Button
                      className="login-button btn-block"
                      color="primary"
                      disabled={isLogging}
                      onClick={handlesOnLogin}
                    >
                      {isLogging ? (
                        <span>
                          login in... &nbsp;
                          <i className="fa fa-spinner fa-pulse fa-fw" />
                        </span>
                      ) : (
                        <span>Login</span>
                      )}
                    </Button>
                  </Col>
                </div>
              </fieldset>
            </form>
          </Col>
        </Row>
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <div className="pull-right">
              <Button className="btn-block" onClick={goHome}>
                back to home
              </Button>
            </div>
          </Col>
        </Row>
      </div>
    </FadeInEntrance>
  );
}

Login.defaultProps = {
  isFetching: false,
  isLogging: false,
};

Login.displayName = 'Login';

export default Login;
