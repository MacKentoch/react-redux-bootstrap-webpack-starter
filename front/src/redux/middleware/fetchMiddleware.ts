import { MiddlewareAPI, Dispatch, Middleware, AnyAction } from 'redux';
import axios from 'axios';

// #region constants
export const FETCH_MOCK = 'FETCH_MOCK';
export const FETCH = 'FETCH';
// #endregion

//
// FETCH_MOCK mode
// in any action just add fetch object like:
// {
//  fetch: {
//    type: 'FETCH_MOCK',
//    actionTypes: {
//      request: 'TYPE_FOR_REQUEST',
//      success: 'TYPE_FOR_RECEIVED',
//      fail: 'TYPE_FOR_ERROR',
//    },
//    mockResult: any
//  }
// }
//

// FETCH mode
// in any action just add fetch object like:
// {
//  fetch: {
//    type: 'FETCH',
//    actionTypes: {
//      request: 'TYPE_FOR_REQUEST',
//      success: 'TYPE_FOR_RECEIVED',
//      fail: 'TYPE_FOR_ERROR',
//    },
//    url: 'an url',
//    method: 'get',  // lower case, one of 'get', 'post'...
//    headers: {}     // OPTIONAL CONTENT like: data: { someprop: 'value ...}
//    options: {}     // OPTIONAL CONTENT like: Authorization: 'Bearer _A_TOKEN_'
//  }
// }

export enum FETCH_TYPE_ENUM {
  FETCH = 'FETCH',
  FETCH_MOCK = 'FETCH_MOCK',
}

export type FetchMiddleWareAction = {
  fetch: {
    type: FETCH_TYPE_ENUM;
    actionTypes?: {
      request: string;
      success: string;
      fail: string;
    };
    url: string;
    method: 'get' | 'put' | 'post' | 'delete';
    headers?: any; // OPTIONAL CONTENT like: data: { someprop: 'value ...}
    options?: any; // OPTIONAL CONTENT like: Authorization: 'Bearer _A_TOKEN_'
  };
  mockResult?: any;
};

const fetchMiddleware: Middleware<Dispatch> = (store: MiddlewareAPI) => (
  next: Function,
) => async (action: AnyAction | FetchMiddleWareAction) => {
  // #region validation
  if (!action.fetch) {
    return next(action);
  }

  if (!action.fetch.type) {
    return next(action);
  }

  // if (
  //   action.fetch.type !== FETCH_TYPE_ENUM.FETCH ||
  //   action.fetch.type !== FETCH_TYPE_ENUM.FETCH_MOCK
  // ) {
  //   return next(action);
  // }

  if (!action.fetch.actionTypes) {
    return next(action);
  }
  // #endregion

  // #region fetch mock
  if (action.fetch.type === FETCH_TYPE_ENUM.FETCH_MOCK) {
    if (!action.fetch.mockResult) {
      throw new Error(
        'Fetch middleware require a mockResult payload when type is "FETCH_MOCK"',
      );
    }

    const {
      actionTypes: { request, success },
      mockResult,
    } = action.fetch;

    // request
    store.dispatch({ type: request });

    // received successful for mock
    return Promise.resolve(
      store.dispatch({
        type: success,
        payload: {
          status: 200,
          data: mockResult,
        },
      }),
    );
  }
  // #endregion

  // #region real fetch
  if (action.fetch.type === FETCH_TYPE_ENUM.FETCH) {
    const {
      actionTypes: { request, success, fail },
      url,
      method,
      headers,
      options,
    } = action.fetch;

    // request
    store.dispatch({ type: request });

    // fetch server (success or fail)
    try {
      const data = await axios.request({
        method,
        url,
        withCredentials: true,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          'Acces-Control-Allow-Origin': '*',
          ...headers,
        },
        ...options,
      });
      store.dispatch({ type: success, payload: data });
      return data;
    } catch (error) {
      store.dispatch({ type: fail, error: error.response });
      throw error;
    }
  }
  // #endregion

  return next(action);
};

export default fetchMiddleware;
