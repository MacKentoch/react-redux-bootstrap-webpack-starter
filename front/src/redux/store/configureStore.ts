import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { createBrowserHistory as createHistory } from 'history';
import { persistReducer, persistStore } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import reducer from '../modules/reducers';

// #region constants
const isProd = process.env.NODE_ENV === 'production';
export const history = createHistory();
// #endregion

// #region createStore : enhancer

// #region logger middleware (dev only)
const loggerMiddleware = createLogger({
  level: 'info',
  collapsed: true,
});
// #endregion

const enhancer = !isProd
  ? composeWithDevTools(
      applyMiddleware(
        thunkMiddleware,
        loggerMiddleware, // logger at the end
      ),
    )
  : composeWithDevTools(applyMiddleware(thunkMiddleware));
// #endregion

// #region persisted reducer
const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['router'],
  // whitelist: ['userAuth'],
};

const persistedReducer = persistReducer(persistConfig, reducer);
// #endregion

export default function configureStore(initialState = {}) {
  const store = createStore(persistedReducer, initialState, enhancer);
  const persistor = persistStore(store);

  if ((module as any).hot) {
    (module as any).hot?.accept('../modules/reducers', () => {
      // @ts-ignore
      const nextAppReducer = require('../modules/reducers').default;
      store.replaceReducer(persistReducer(persistConfig, nextAppReducer));
    });
  }

  return { store, persistor };
}
