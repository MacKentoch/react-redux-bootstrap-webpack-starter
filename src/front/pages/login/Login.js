// @flow

// #region imports
import React, { PureComponent } from 'react';
import {
  type Match,
  type Location,
  type RouterHistory,
} from 'react-router-dom';
import Button from 'reactstrap/lib/Button';
import Row from 'reactstrap/lib/Row';
import Col from 'reactstrap/lib/Col';
import auth from '../../services/auth';
import { type UserAuthActions } from '../../types/redux/userAuth';
// #endregion

// #region flow types
export type Props = {
  // react-router 4:
  match: Match,
  location: Location,
  history: RouterHistory,

  // userAuth:
  isAuthenticated: boolean,
  isFetching: boolean,
  isLogging: boolean,
} & UserAuthActions;

export type State = {
  email: string,
  password: string,
};
// #endregion

class Login extends PureComponent<Props, State> {
  static defaultProps = {
    isFetching: false,
    isLogging: false,
  };

  state = {
    email: '',
    password: '',
  };

  // #region lifecycle methods
  componentDidMount() {
    const { disconnectUser } = this.props;

    disconnectUser(); // diconnect user: remove token and user info
  }

  render() {
    const { email, password } = this.state;

    const { isLogging } = this.props;

    return (
      <div className="content">
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <form className="form-horizontal" noValidate>
              <fieldset>
                <legend>Login</legend>

                <div className="form-group">
                  <label
                    htmlFor="inputEmail"
                    className="col-lg-2 control-label"
                  >
                    Email
                  </label>
                  <Col lg={12}>
                    <input
                      type="text"
                      className="form-control"
                      id="inputEmail"
                      placeholder="Email"
                      autoComplete="username email"
                      value={email}
                      onChange={this.handlesOnEmailChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <label
                    htmlFor="inputPassword"
                    className="col-lg-2 control-label"
                  >
                    Password
                  </label>
                  <Col lg={12}>
                    <input
                      type="password"
                      className="form-control"
                      id="inputPassword"
                      placeholder="Password"
                      autoComplete="current-password"
                      value={password}
                      onChange={this.handlesOnPasswordChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <Col lg={{ size: 12 }}>
                    <Button
                      className="login-button btn-block"
                      color="primary"
                      disabled={isLogging}
                      onClick={this.handlesOnLogin}
                    >
                      {isLogging ? (
                        <span>
                          login in... &nbsp;
                          <i className="fa fa-spinner fa-pulse fa-fw" />
                        </span>
                      ) : (
                        <span>Login</span>
                      )}
                    </Button>
                  </Col>
                </div>
              </fieldset>
            </form>
          </Col>
        </Row>
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <div className="pull-right">
              <Button className="btn-block" onClick={this.goHome}>
                back to home
              </Button>
            </div>
          </Col>
        </Row>
      </div>
    );
  }
  // #endregion

  // #region form inputs change callbacks
  handlesOnEmailChange = (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
      // should add some validator before setState in real use cases
      this.setState({ email: event.target.value.trim() });
    }
  };

  handlesOnPasswordChange = (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
      // should add some validator before setState in real use cases
      this.setState({ password: event.target.value.trim() });
    }
  };
  // #endregion

  // #region on login button click callback
  handlesOnLogin = async (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
    }

    const { history, logUserIfNeeded } = this.props;

    const { email, password } = this.state;

    const userLogin = {
      login: email,
      password: password,
    };

    try {
      const response = await logUserIfNeeded(userLogin);
      const {
        data: { token, user },
      } = response.payload;

      auth.setToken(token);
      auth.setUserInfo(user);

      history.push({ pathname: '/' }); // back to Home
    } catch (error) {
      /* eslint-disable no-console */
      console.log('login went wrong..., error: ', error);
      /* eslint-enable no-console */
    }
  };
  // #endregion

  // #region on go back home button click callback
  goHome = (event: SyntheticEvent<>) => {
    if (event) {
      event.preventDefault();
    }

    const { history } = this.props;

    history.push({ pathname: '/' });
  };
  // #endregion
}

export default Login;
