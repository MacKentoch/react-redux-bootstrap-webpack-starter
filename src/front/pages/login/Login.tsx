import React, { useState, useEffect } from 'react';
import { RouteComponentProps } from 'react-router';
import Button from 'reactstrap/lib/Button';
import Row from 'reactstrap/lib/Row';
import Col from 'reactstrap/lib/Col';
import auth from '../../services/auth';
import FadeInEntrance from '../../components/fadeInEntrance';
import { MappedDispatchToProps, MappedStateToProps, OwnProps } from './index';

// #region types
export type Props = {} & RouteComponentProps &
  MappedStateToProps &
  MappedDispatchToProps &
  OwnProps;
// #endregion

function Login({
  isLogging,
  isFetching,
  disconnectUser,
  logUserIfNeeded,
  history,
}: Props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  useEffect(() => {
    // on mount disconnect (clear previous user auth and info):
    disconnectUser();
  }, []);

  // #region form inputs change callbacks
  const handlesOnEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event) {
      event.preventDefault();
      setEmail(event.target.value.trim());
    }
  };

  const handlesOnPasswordChange = (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    if (event) {
      event.preventDefault();
      setPassword(event.target.value.trim());
    }
  };
  // #endregion

  // #region on login button click callback
  const handlesOnLogin = async (event: React.MouseEvent<HTMLButtonElement>) => {
    event && event.preventDefault();

    try {
      const response = await logUserIfNeeded(email, password);
      console.log('"response: ', response);
      const {
        data: { token, user },
      } = response.payload;

      auth.setToken(token);
      auth.setUserInfo(user);

      history.push({ pathname: '/' }); // back to Home
    } catch (error) {
      /* eslint-disable no-console */
      console.log('login went wrong..., error: ', error);
      /* eslint-enable no-console */
    }
  };
  // #endregion

  // #region on go back home button click callback
  const goHome = (event: React.MouseEvent<HTMLButtonElement>) => {
    event && event.preventDefault();
    history.push({ pathname: '/' });
  };
  // #endregion

  return (
    <FadeInEntrance>
      <div className="content">
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <form className="form-horizontal" noValidate>
              <fieldset>
                <legend>Login</legend>
                <div className="form-group">
                  <label
                    htmlFor="inputEmail"
                    className="col-lg-2 control-label"
                  >
                    Email
                  </label>
                  <Col lg={12}>
                    <input
                      type="text"
                      className="form-control"
                      id="inputEmail"
                      placeholder="Email"
                      autoComplete="username email"
                      value={email}
                      onChange={handlesOnEmailChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <label
                    htmlFor="inputPassword"
                    className="col-lg-2 control-label"
                  >
                    Password
                  </label>
                  <Col lg={12}>
                    <input
                      type="password"
                      className="form-control"
                      id="inputPassword"
                      placeholder="Password"
                      autoComplete="current-password"
                      value={password}
                      onChange={handlesOnPasswordChange}
                    />
                  </Col>
                </div>

                <div className="form-group">
                  <Col lg={{ size: 12 }}>
                    <Button
                      className="login-button btn-block"
                      color="primary"
                      disabled={isLogging}
                      onClick={handlesOnLogin}
                    >
                      {isLogging ? (
                        <span>
                          login in... &nbsp;
                          <i className="fa fa-spinner fa-pulse fa-fw" />
                        </span>
                      ) : (
                        <span>Login</span>
                      )}
                    </Button>
                  </Col>
                </div>
              </fieldset>
            </form>
          </Col>
        </Row>
        <Row>
          <Col md={{ size: 4, offset: 4 }} xs={{ size: 10, offset: 1 }}>
            <div className="pull-right">
              <Button className="btn-block" onClick={goHome}>
                back to home
              </Button>
            </div>
          </Col>
        </Row>
      </div>
    </FadeInEntrance>
  );
}

Login.defaultProps = {
  isFetching: false,
  isLogging: false,
};

Login.displayName = 'Login';

export default Login;
