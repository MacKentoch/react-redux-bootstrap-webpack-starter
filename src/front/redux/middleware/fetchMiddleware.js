// @flow

// #region imports
import axios from 'axios';
// #endregion

// #region constants
export const FETCH_MOCK = 'FETCH_MOCK';
export const FETCH = 'FETCH';
// #endregion

//
// FETCH_MOCK mode
// in any action just add fetch object like:
// {
//  fetch: {
//    type: 'FETCH_MOCK',
//    actionTypes: {
//      request: 'TYPE_FOR_REQUEST',
//      success: 'TYPE_FOR_RECEIVED',
//      fail: 'TYPE_FOR_ERROR',
//    },
//    mockResult: any
//  }
// }
//

// FETCH mode
// in any action just add fetch object like:
// {
//  fetch: {
//    type: 'FETCH',
//    actionTypes: {
//      request: 'TYPE_FOR_REQUEST',
//      success: 'TYPE_FOR_RECEIVED',
//      fail: 'TYPE_FOR_ERROR',
//    },
//    url: 'an url',
//    method: 'get',  // lower case, one of 'get', 'post'...
//    headers: {}     // OPTIONAL CONTENT like: data: { someprop: 'value ...}
//    options: {}     // OPTIONAL CONTENT like: Authorization: 'Bearer _A_TOKEN_'
//  }
// }
//
//
//
//
const fetchMiddleware = store => next => action => {
  if (!action.fetch) {
    return next(action);
  }

  if (
    !action.fetch.type ||
    !action.fetch.type === FETCH_MOCK ||
    !action.fetch.type === FETCH
  ) {
    return next(action);
  }

  if (!action.fetch.actionTypes) {
    return next(action);
  }

  /**
   * fetch mock
   * @type {[type]}
   */
  if (action.fetch.type === FETCH_MOCK) {
    if (!action.fetch.mockResult) {
      throw new Error(
        'Fetch middleware require a mockResult payload when type is "FETCH_MOCK"',
      );
    }

    const {
      actionTypes: { request, success },
      mockResult,
    } = action.fetch;

    // request
    store.dispatch({ type: request });

    // received successful for mock
    return Promise.resolve(
      store.dispatch({
        type: success,
        payload: {
          status: 200,
          data: mockResult,
        },
      }),
    );
  }

  if (action.fetch.type === FETCH) {
    const {
      actionTypes: { request, success, fail },
      url,
      method,
      headers,
      options,
    } = action.fetch;

    // request
    store.dispatch({ type: request });

    // fetch server (success or fail)
    // returns a Promise
    return axios
      .request({
        method,
        url,
        withCredentials: true,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          'Acces-Control-Allow-Origin': '*',
          ...headers,
        },
        ...options,
      })
      .then(data => store.dispatch({ type: success, payload: data }))
      .catch(err => {
        store.dispatch({ type: fail, error: err.response });
        return Promise.reject(err.response);
      });
  }
  return next(action);
};

export default fetchMiddleware;
