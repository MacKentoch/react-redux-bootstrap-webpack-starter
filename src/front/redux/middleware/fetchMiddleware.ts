import { MiddlewareAPI, Dispatch, Middleware, AnyAction } from 'redux';
import axios from 'axios';

// #region constants
export const FETCH_MOCK = 'FETCH_MOCK';
export const FETCH = 'FETCH';
// #endregion

//
// FETCH_MOCK mode
// in any action just add fetch object like:
// {
//  fetch: {
//    type: 'FETCH_MOCK',
//    actionTypes: {
//      request: 'TYPE_FOR_REQUEST',
//      success: 'TYPE_FOR_RECEIVED',
//      fail: 'TYPE_FOR_ERROR',
//    },
//    mockResult: any
//  }
// }
//

// FETCH mode
// in any action just add fetch object like:
// {
//  fetch: {
//    type: 'FETCH',
//    actionTypes: {
//      request: 'TYPE_FOR_REQUEST',
//      success: 'TYPE_FOR_RECEIVED',
//      fail: 'TYPE_FOR_ERROR',
//    },
//    url: 'an url',
//    method: 'get',  // lower case, one of 'get', 'post'...
//    headers: {}     // OPTIONAL CONTENT like: data: { someprop: 'value ...}
//    options: {}     // OPTIONAL CONTENT like: Authorization: 'Bearer _A_TOKEN_'
//  }
// }

export enum FETCH_TYPE_ENUM {
  'FETCH', 'FETCH_MOCK'
}

export type FetchMiddleWareAction = {
   fetch: {
     type: FETCH_TYPE_ENUM,
     actionTypes?: {
       request: string,
       success: string,
       fail: string,
     },
     url: string,
     method: 'get' | 'put' | 'post' | 'delete',
     headers?: any     // OPTIONAL CONTENT like: data: { someprop: 'value ...}
     options?: any     // OPTIONAL CONTENT like: Authorization: 'Bearer _A_TOKEN_'
   },
   mockResult?: any
}

const fetchMiddleware: Middleware<Dispatch> = (store: MiddlewareAPI) => (
  next: Function,
) => (action: AnyAction | FetchMiddleWareAction) => {
  if (!action.fetch) {
    return next(action);
  }

  if (
    !action.fetch.type ||
    !action.fetch.type === FETCH_TYPE_ENUM.FETCH_MOCK ||
    !action.fetch.type === FETCH_TYPE_ENUM.FETCH
  ) {
    return next(action);
  }

  if (!action.fetch.actionTypes) {
    return next(action);
  }

  /**
   * fetch mock
   * @type {[type]}
   */
  if (action.fetch.type === FETCH_MOCK) {
    if (!action.fetch.mockResult) {
      throw new Error(
        'Fetch middleware require a mockResult payload when type is "FETCH_MOCK"',
      );
    }

    const {
      actionTypes: { request, success },
      mockResult,
    } = action.fetch;

    // request
    store.dispatch({ type: request });

    // received successful for mock
    return Promise.resolve(
      store.dispatch({
        type: success,
        payload: {
          status: 200,
          data: mockResult,
        },
      }),
    );
  }

  if (action.fetch.type === FETCH) {
    const {
      actionTypes: { request, success, fail },
      url,
      method,
      headers,
      options,
    } = action.fetch;

    // request
    store.dispatch({ type: request });

    // fetch server (success or fail)
    // returns a Promise
    return axios
      .request({
        method,
        url,
        withCredentials: true,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          'Acces-Control-Allow-Origin': '*',
          ...headers,
        },
        ...options,
      })
      .then(data => store.dispatch({ type: success, payload: data }))
      .catch(err => {
        store.dispatch({ type: fail, error: err.response });
        return Promise.reject(err.response);
      });
  }
  return next(action);
};

export default fetchMiddleware;
