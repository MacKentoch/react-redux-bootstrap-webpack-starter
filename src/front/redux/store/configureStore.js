// @flow

/* eslint no-process-env:0 */
if (process.env.NODE_ENV === 'production') {
  module.exports = require('./configureStore.prod');
} else {
  module.exports = require('./configureStore.dev');
}
