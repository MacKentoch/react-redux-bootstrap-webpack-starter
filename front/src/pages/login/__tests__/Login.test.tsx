import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import { ThemeProvider } from 'styled-components';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render } from '@testing-library/react';
import fetchMiddleware from '../../../redux/middleware/fetchMiddleware';
import Login from '../index';

const middlewares = [thunk, fetchMiddleware];
const mockStore = configureStore(middlewares);

describe('Login page', () => {
  let rootElement: any = null;

  beforeEach(() => {
    rootElement = document.createElement('div');
    document.body.appendChild(rootElement);
  });

  afterEach(() => {
    rootElement && document.body.removeChild(rootElement);
    rootElement = null;
  });

  it('renders as expected', () => {
    const initialState = {
      userAuth: {
        isFetching: false,
        isLogging: false,
        id: '',
        login: '',
        firstname: '',
        lastname: '',
        token: '',
        isAuthenticated: false,
      },
    };
    const store = mockStore(initialState);

    const { container } = render(
      <Provider store={store}>
        <ThemeProvider theme={{}}>
          <MemoryRouter>
            <Login />
          </MemoryRouter>
        </ThemeProvider>
      </Provider>,
      rootElement,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
