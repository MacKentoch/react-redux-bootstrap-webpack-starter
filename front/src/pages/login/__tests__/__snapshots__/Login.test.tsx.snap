// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders as expected 1`] = `
<div
  class="sc-bdVaJa jxVqOJ"
>
  <div
    class="content"
  >
    <div
      class="row"
    >
      <div
        class="col-10 offset-1 col-md-4 offset-md-4"
      >
        <form
          class="form-horizontal"
          novalidate=""
        >
          <fieldset>
            <legend>
              Login
            </legend>
            <div
              class="form-group"
            >
              <label
                class="col-lg-2 control-label"
                for="inputEmail"
              >
                Email
              </label>
              <div
                class="col-lg-12"
              >
                <input
                  autocomplete="username email"
                  class="form-control"
                  id="inputEmail"
                  placeholder="Email"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="form-group"
            >
              <label
                class="col-lg-2 control-label"
                for="inputPassword"
              >
                Password
              </label>
              <div
                class="col-lg-12"
              >
                <input
                  autocomplete="current-password"
                  class="form-control"
                  id="inputPassword"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="form-group"
            >
              <div
                class="col-lg-12"
              >
                <button
                  class="login-button btn-block btn btn-primary"
                  type="button"
                >
                  <span>
                    Login
                  </span>
                </button>
              </div>
            </div>
          </fieldset>
        </form>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-10 offset-1 col-md-4 offset-md-4"
      >
        <div
          class="pull-right"
        >
          <button
            class="btn-block btn btn-secondary"
            type="button"
          >
            back to home
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
